data_model/src/lib.rs

```diff
@@ -1821,6 +1818,21 @@ pub mod peer {
         pub public_key: PublicKey,
     }
 
+    impl PartialEq for Id {
+        fn eq(&self, other: &Self) -> bool {
+            // Comparison is done by public key only, so that full domain names can be present in trusted peers and local address in this peer config without a conflict.
```

Comment:
I think it would be better to place the comment on the type declaration? Also use a line break :)

can you make the comment even more descriptive? I don't get it really and this piece of code is contentious so I think a thorough explanation is worth it here. I have the feeling that user should uphold the invariant that each peer is assigned different public key. What would happen in that scenario or is this scenario considered UB?

Reply:
If there are invariants to watch for, they should be listed in documentation of the type

Reply:
Yes, you are correct, each peer should have a unique public key. I will add this to the docs

Reply:
And about your question - this is needed so that we can have something like `peer1.some.domain.com` in trusted_peers, but the peer itself will be started at `127.0.0.1:1337`, or `0.0.0.0:1337`, etc.

Then a peer will can identify itself in the trusted peers without resolving the IP based on domain name.

Reply:
Ah yes, I've looked at the issue later. Add that to the comment please. 

So we can either:
1. define an invariant that each peer should have a unique public key(in this case it's UB to assign the same public key to two peers)
2. define what happens if two peers are assigned the same public key
3. do domain name resolution before storing into peer id

Reply:
If we go with 1. we can/should? also make some wrapper struct which will enforce this invariant for us on the collection of peer ids. Something like `SignaturesOf`

Reply:
I lean towards Option 2. If two peers are assigned the same public key, they could be treated as entities that have a short-circuit connection, i.e. they don't need to gossip in order to exchange information. 
Option 3 is also IMO quite acceptable. 
Option 1 is going to be extremely hard to enforce, and not the kind of fragility that I'm personally comfortable with. Although the same can be said about options 2 and 3.

Reply:
Fragments of thoughts:

* If PK is unique to peer, peer can be identified by PK


* If incoming, a peer has a resolved address
* An unresolved address is for outgoing use

Currently the network control flow is:
* Domestic registrations -> WSV whitelist of peers -> p2p connections

This order would have to be a bit reversed in public chains, where incomings to p2p should be capable of affecting WSV

Reply:
If PK is unique to peer, peer can be identified by PK.

Regarding peer address, my understanding is that
why we have `String` instead of `SocketAddr` is just for the SCALE Codec.
Any peer address WSV has should be successfully parsed to a resolved address

Reply:
Egor Ivkov, [11 Jan 2022 14:28:46]
mtakemiya can you comment on this? https://github.com/hyperledger/iroha/pull/1800#discussion_r781217943

I think initially we wanted for each peer to have a unique public key, right?

I don’t ask for crypto. M◉⃤|<𓊖༒𓊖 Takemiya 武宮誠 (Sora.org - SORAMITSU), [11 Jan 2022 20:24:08]
[In reply to Egor Ivkov]
Yes, of course each peer should have a unique public key

Reply:
I am going to proceed and create an issue to report misconfiguration error if at least two identical public keys are found in trusted peers.

Reply:
https://github.com/hyperledger/iroha/issues/1811